crates/go/src/wit_future.go

```diff
@@ -0,0 +1,131 @@
+package wit_types
+
+import (
+	"runtime"
+	"unsafe"
+	"wit_component/wit_async"
+	"wit_component/wit_runtime"
+)
+
+type FutureVtable[T any] struct {
+	Size         uint32
+	Align        uint32
+	Read         func(handle int32, item unsafe.Pointer) uint32
+	Write        func(handle int32, item unsafe.Pointer) uint32
+	CancelRead   func(handle int32) uint32
+	CancelWrite  func(handle int32) uint32
+	DropReadable func(handle int32)
+	DropWritable func(handle int32)
+	Lift         func(src unsafe.Pointer) T
+	Lower        func(pinner *runtime.Pinner, value T, dst unsafe.Pointer)
+}
+
+type FutureReader[T any] struct {
+	vtable *FutureVtable[T]
+	handle int32
+}
+
+func (f *FutureReader[T]) Read() T {
+	if f.handle == 0 {
+		panic("null future handle")
+	}
+
+	handle := f.handle
+	f.handle = 0
+	defer f.vtable.DropReadable(handle)
+
+	pinner := runtime.Pinner{}
+	defer pinner.Unpin()
+
+	buffer := wit_runtime.Allocate(&pinner, uintptr(f.vtable.Size), uintptr(f.vtable.Align))
+
+	code, _ := wit_async.FutureOrStreamWait(f.vtable.Read(handle, buffer), handle)
+
+	switch code {
+	case wit_async.RETURN_CODE_COMPLETED, wit_async.RETURN_CODE_DROPPED:
```

Comment:
`RETURN_CODE_DROPPED` here should be a panic, right? (it shouldn't be possible anyway to drop the writer without writing)